### PR TITLE
fix(tui): place a `pane spawn`ed role at its configured index, not at the end (#1096)

### LINKED ISSUE #1096: `pane spawn`: a role inserted mid-config shifts role indices and breaks orchestration restore
## What is wrong

`TabManager::add_role_to_existing_orchestration` (`src/tab.rs:1334`, added by PR #918 / issue #868) grows an already-open orchestration tab when `pane spawn <role>` brings up a role that was not running. A role absent from the open tab is **always appended**, regardless of where it sits in `.dot-agent-deck.toml`.

Concrete failure — changing a config from:

```toml
roles = [orchestrator, coder]
```

to:

```toml
roles = [orchestrator, reviewer, coder]
```

then running `dot-agent-deck pane spawn reviewer` leaves the live tab and the persisted session snapshot ordered `[orchestrator, coder, reviewer]`. `resolve_orchestration_for_restore` compares that exact sequence against the current config, sees drift, rejects the match, and **falls back to restoring the orchestration as a plain pane** — the user loses the orchestration tab on the next restore.

## Status: known, documented, accepted for M4 — but untracked until now

This is not an oversight. It was raised by Greptile on PR #918 as a P1, and resolved as deliberately out of scope. The function's own doc comment says so:

> A role inserted in the MIDDLE of the config, shifting existing role indices, is a known out-of-scope edge case for M4 — not handled here.

and `tests/CATALOG.md` records the same boundary under `pane/spawn/009`:

> **Does not assert:** a role inserted in the MIDDLE of the config, shifting existing role indices (a known out-of-scope edge case for M4).

The append and replace paths that M4 *does* claim are covered by `pane/spawn/009` and `pane/spawn/010`.

The resolution comment on that thread said the case is *"real and worth its own issue if we want it actually closed rather than documented"* — no such issue was filed. This is it. Filed while verifying #918, after searching the open issues and finding none.

## What closing it would take

Two shapes, per Greptile's own framing:

1. **Preserve the configured role index.** Look the spawned role's index up in the current `.dot-agent-deck.toml` and insert at that position rather than appending, shifting `role_pane_ids` / `role_statuses` / `config.roles` together so all three stay parallel. Also needs the persisted snapshot in `ui.pane_metadata` to be rebuilt in config order rather than pushed onto — `src/ui.rs`'s growth path currently does `snapshot.roles.push(...)` on a genuine append.
2. **Explicitly reject non-append additions.** Have `add_role_to_existing_orchestration` return a `TabError` when the role's config index is not the next slot, so `pane spawn` fails loudly with a "restart the tab to pick up a reordered config" message instead of silently producing a tab that will not restore.

(2) is much cheaper and arguably the honest M4 behaviour; (1) is what actually closes the user-facing gap. Either needs a test pinning the mid-insert case — `pane/spawn/009`'s "Does not assert" line is the slot it goes in.

Worth checking while in there: the `role_index` the daemon stamps on `TabMembership` comes from `lookup_orchestration_role_indexed` against the *current* config, while the tab's own `config.roles` is whatever it was opened with. A mid-config insert makes those two disagree, which is the same root cause seen from the daemon side.

## Depends on

#918 landing first — `add_role_to_existing_orchestration` does not exist on `main` yet.

### PR DESCRIPTION
**Pull request by vfarcic**
Closes #1096.

## The bug

`TabManager::add_role_to_existing_orchestration` (`src/tab.rs`) grows an already-open orchestration tab when `pane spawn <role>` brings up a role that was not running. A role absent from the tab was **always appended**, wherever `.dot-agent-deck.toml` actually put it.

The issue's worked case: edit `roles = [orchestrator, coder]` into `[orchestrator, reviewer, coder]`, run `dot-agent-deck pane spawn reviewer`, and both the live tab and the persisted session snapshot come out ordered `[orchestrator, coder, reviewer]`.

That breaks restore rather than merely looking wrong. `resolve_orchestration_for_restore` compares the saved role sequence against the config **element by element**, so a snapshot in the wrong order is rejected exactly like one missing a role: drift is reported, the orchestration is rebuilt as a plain pane, and the tab is gone. The `role_index` the daemon stamps on `TabMembership` comes from the *current* config, so it disagreed with the tab's own order at the same moment — the issue's "same root cause seen from the daemon side".

This was known and documented rather than missed: PR #918 / issue #868 resolved it as out of scope for M4, in the function's own doc comment and in `pane/spawn/009`'s **Does not assert** line. This is issue #1096's "actually closed rather than documented".

## The fix

The issue offered two shapes. This is shape **(1)** — preserve the configured index — because (2) (reject a non-append addition with a "restart the tab" error) leaves the user-facing gap open.

- **`src/tab.rs`** — `add_role_to_existing_orchestration` takes the current config's role list and **places** the role via a new `role_placement_index`: insert before the first tab role the current config puts *after* the new one, append when there is none. `role_pane_ids`, `role_statuses` and the tab's `config.roles` shift together so they stay parallel, and `start_role_index` moves with them — so `role_pane_ids[start_role_index]` still names the orchestrator's pane, which is the `ui.pane_metadata` key the snapshot hangs off and the pane the orchestrator prompt is delivered to. The role-index cap check is unchanged: the highest index this growth produces is the old length wherever the slot lands.
- **`src/ui.rs`** — the growth path passes `&orch_config.roles` (the freshly-loaded current config it already resolved `role_config` from), and **rebuilds** the snapshot's role list from the tab instead of `snapshot.roles.push(...)`. The tab is the source of truth for display order — `OrchestrationSnapshot.roles` is documented as "the same order as the tab's `role_pane_ids`" — and a push would now record an order the tab does not have. Rebuilding is also a no-op on the replace path, which is what the old `if was_new` guard bought. `started_role_indices` (forward-compat, no reader today) is shifted so it keeps pointing at the same roles.

The issue's second bullet is the reason `src/ui.rs` is touched at all; it names that file's `snapshot.roles.push(...)` as part of the fix. Nothing else outside `src/tab.rs` moved.

## What is NOT reconciled, and why

Only a role being **added** is repaired. Named explicitly rather than left implicit:

| config drift since the tab opened | handled? |
| --- | --- |
| role **inserted** mid-list | **yes** — placed at its configured index |
| role **appended** at the end | **yes** — unchanged behaviour, pinned by `pane/spawn/015` |
| role **removed** | no |
| two roles **reordered** | no |
| role **renamed** | no (reads as a remove + an add) |

None of the last three is expressible as a `pane spawn` of one role, and each would mean deleting or moving slots whose panes are **live agents** — a bigger promise than this verb makes. Where such a drift leaves the tab unable to express the current config at all, placement degrades to the old append and restore falls back to a plain pane exactly as it does today. A role the current config does not list at all appends for the same reason.

## Tests

Reproduce-first: `pane/spawn/014` was written against the pre-fix code and watched fail for the user's reason — `left: ["orchestrator", "coder", "reviewer"]`, `right: ["orchestrator", "reviewer", "coder"]` — before anything was fixed.

- **`pane/spawn/014`** (new, L1, `TabManager` only) — the issue's exact case: a role inserted mid-config lands at index 1, all three parallel vectors in lockstep, returned index 1 (the daemon's index, not the append index 2), `start_role_index` unmoved.
- **`pane/spawn/015`** (new, L1) — the control. A role configured **last** still appends, and a role the current config does not mention at all still appends rather than guessing a slot or panicking.
- **`pane/spawn/016`** (new, L1) — inserting before the start role carries `start_role_index` with it (1 → 2), so `role_pane_ids[start_role_index]` still names the orchestrator's pane. Reverting only that half turns this red on its own.
- **`pane/drift/001`** (extended, L2 PTY) — its fixture now **inserts** `reviewer` between `orchestrator` and `coder` instead of appending it. An end-of-list role lands in the right slot either way, so the old fixture could not tell placement from appending. The test parses the flushed `session.toml` rather than substring-matching it (the drift guard is about order, and the TOML writer emits a multi-line array) and asserts `[orchestrator, reviewer, coder]`. Reverting the `src/ui.rs` rebuild alone turns it red.
- **`pane/spawn/005`, `pane/spawn/009`, `pane/spawn/010`** — updated for the new argument and the changed fixture; `009`'s stale **Does not assert** line now points at `014`.

Ran locally:

- `cargo test-fast` — **3152 passed, 0 failed** (41.0s).
- `cargo test-e2e --test e2e_pane_spawn_live` — 2 passed (`spawn_005`, `drift_001`).
- `cargo test-e2e --test e2e_session_save --test e2e_session_restore` — 15 passed (the snapshot-write path this touches).
- `cargo nextest run --workspace --test orchestration_tab_growth` — 5 passed.
- `cargo fmt --check`, `cargo clippy --workspace --all-targets --features e2e,e2e-live -- -D warnings`, `cargo xtask linkage-check` — all clean.

No lane-2 test is involved: nothing here reaches a real agent.

## Rule 12

`PROTOCOL_VERSION` is **9** on `main` (verified in `src/daemon_protocol.rs`) and stays 9. Nothing on the TUI↔daemon wire moved and no field changed meaning: `OrchestrationSurface` already carried each role's `role_index` resolved against the current config — this change is the TUI finally *honouring* that in its own tab bookkeeping, not the daemon saying something new. The one persisted shape that changes, `[panes.orchestration]`'s `roles` order, is written and read by the same TUI, and an old TUI reading a new snapshot sees a role list in config order, which is the order its own drift guard wants.

**The cross-version manual test was run** (this PR originally claimed it did not apply; the automated review was right that it does, and the "no wire change" reasoning is exactly what that rule overrides). Isolated sandbox, own `HOME`/sockets/state/log/flags, `DOT_AGENT_DECK_IDLE_SHUTDOWN_SECS=0`. Sequence: `v0.40.2` daemon (`0.40.2-g7e87d7f`) → `v0.40.2` TUI to bring both roles up under it → detach → branch TUI (`0.40.2-g3872d0e9`), mismatch prompt naming both live agents **declined**.

| flow, against the older daemon | result |
| --- | --- |
| `delegate --to coder` | routes — `worker-task-coder.md` written, pointer rendered in the coder's pane |
| `work-done` hook | arrives — delegation retired, `work-done-coder.md` written |
| `agent-event --type running` | arrives — card flipped to `Thinking` (done **last**; it reclassifies the pane as `Pi`) |

Confirmed genuine rather than a silently restarted same-version test: exactly **one** `Attach protocol listening` line for the whole run, no idle-shutdown line, same daemon pid with the v0.40.2 binary still as its `/proc/<pid>/exe` throughout. Both sides at protocol 9, so it exercised semantics behind a stable wire. The change itself was exercised there too — `pane spawn reviewer` against the older daemon placed `reviewer` between `orchestrator` and `coder`. Teardown was by the captured pid after re-reading its `exe`, never a `pkill` pattern.

That run does not reach the snapshot half: a tab rebuilt by attaching to a running daemon captures no `ui.pane_metadata`, so there is no entry to update on that path. Pre-existing and already recorded as out of scope in the growth code's own comment; `pane/drift/001` covers the snapshot half on the form-opened path. Nothing silently stopped flowing, so no `PROTOCOL_VERSION` bump and no `.breaking.md`; `changelog.d/1096.bugfix.md` covers it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/e2e_pane_spawn_live.rs [hunk 1]
```diff
@@ -256,25 +286,29 @@ fn spawn_005_pane_spawn_joins_the_already_open_orchestration_tab() {
...
 /// later snapshot flush keeps re-serializing the ORIGINAL two-role list. A
 /// restored session would then see `resolve_orchestration_for_restore`'s
 /// drift guard false-positive (`current_roles` re-read from the now-3-role
 /// `.dot-agent-deck.toml` vs. `saved_roles` frozen at 2) and fall back to a
 /// plain pane, discarding the whole orchestration tab reconstruction even
-/// though nothing on disk ever actually diverged.
+/// though nothing on disk ever actually diverged. Issue #1096 adds the ORDER
+/// half: the guard is an element-by-element sequence comparison, so a
+/// snapshot that carries `reviewer` but appended it is rejected exactly like
+/// one that never captured it.
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply tests/e2e_pane_spawn_live.rs]: Valid — fixed in 72de6715.

CLAUDE.md rule 7 asks for 1–3 sentences and this one ran to eleven. It was already over budget before this PR, but my change lengthened it rather than leaving it alone, so trimming it is mine to do. `pane/spawn/005`'s comment was over too and I had added a clause to it, so both are now three sentences covering setup, action and expected result.

The #868/#1096 rationale is not deleted, just relocated: it sits beside each test as ordinary `//` comments, which keeps it out of the `/// Scenario:` budget and out of the generated `.md` while staying next to the code it explains.

`cargo xtask linkage-check` (rule 7) and `cargo xtask docs --tests` both pass, and the two tests still pass (`cargo test-e2e --test e2e_pane_spawn_live`).

File: changelog.d/1096.bugfix.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+**What it does not reconcile, deliberately.** Only a role being *added* is repaired. A role **removed** from the config, two roles **reordered**, or a role **renamed** are not: none is expressible as a `pane spawn` of one role, and each would mean deleting or moving slots whose panes are live agents — a bigger promise than this verb makes. Where such a drift leaves the tab unable to express the current config at all, placement degrades to the old append and restore falls back to a plain pane exactly as it does today. Two narrower degradations are the same by construction: a role the current config does not list at all appends, and existing slots are never moved.
+
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply changelog.d/1096.bugfix.md:10]: You're right, and the reasoning I used was the exact one rule 12 exists to override. The test has now been run — 72de6715 replaces the "does not apply" claim in the changelog fragment with the run.

Rule 12 requires the cross-version test for any PR touching the daemon, protocol, **orchestration**, or hooks, and that step is not conditional on the "did this change the contract?" answer above it. A semantic break behind a stable wire is precisely what it catches, so "no wire change" is not a reason to skip it. `docs/develop/versioning.md` §"Cross-version manual-test discipline" says the same thing.

**Setup.** Isolated sandbox on a disk-backed path with its own `HOME`, `XDG_*`, hook/attach sockets, `DOT_AGENT_DECK_STATE_DIR`, `DOT_AGENT_DECK_LOG`, `DOT_AGENT_DECK_SESSION` and `DOT_AGENT_DECK_EXPERIMENTAL=0`, plus `DOT_AGENT_DECK_IDLE_SHUTDOWN_SECS=0` so the 30s idle window could not silently swap the daemon.

**Sequence.** `v0.40.2` daemon (build id `0.40.2-g7e87d7f`) → `v0.40.2` TUI to bring both roles of a two-role orchestration up under it → detach that TUI (agents survive, confirmed via `daemon status`) → branch TUI (`0.40.2-g3872d0e9`), whose build-mismatch prompt named both live agents, **declined**.

**Results against the older daemon.**

| flow | result |
| --- | --- |
| `delegate --to coder` | routes — daemon wrote `worker-task-coder.md`, its pointer rendered in the coder's pane |
| `work-done` hook | arrives — delegation retired, `work-done-coder.md` written |
| `agent-event --type running` | arrives — card flipped to `Thinking` (done **last**, since it reclassifies the pane as `Pi`) |

**Confirmed genuine, not a silently restarted same-version test:** exactly **one** `Attach protocol listening` line for the whole run, no idle-shutdown line, and the same daemon pid with the v0.40.2 binary still as its `/proc/<pid>/exe` from start to finish. Both sides at `PROTOCOL_VERSION` 9, so this exercised semantics behind a stable wire rather than a handshake rejection — which is what the rule is for.

I also exercised the change itself in that configuration: `pane spawn reviewer` against the older daemon placed `reviewer` **between** `orchestrator` and `coder`, at its configured index.

Nothing silently stopped flowing, so no `PROTOCOL_VERSION` bump and no `.breaking.md`.

One limit worth stating rather than leaving implicit: that run does not reach the snapshot half, because a tab rebuilt by attaching to a running daemon captures no `ui.pane_metadata` at all, so there is no entry to update on that path. That is pre-existing and already recorded as out of scope in the growth code's own comment — `pane/drift/001` covers the snapshot half on the form-opened path where the metadata does exist.

Teardown was by the captured pid after re-reading its `/proc/<pid>/exe`, never a `pkill` pattern (CLAUDE.md rule 12's teardown bullet / #428 occurrence #5). The box's production daemon was verified still serving afterwards.
